Build: Update calypso-build to v4

sirreal opened the pull request:
Update `@automattic/calypso-build` to v4. Required for #13490

Closes #13594 (the automated dependency upgrade PR). 

This updates the `calypso-build` dependency to version 4.

sirreal committed:
Upgrade calypso build to 4.1.0

sirreal committed:
Apply wordpress-element babel transform in Jest

sirreal commented on tests/jest-extensions-babel-transform.js:
```diff
@@ -0,0 +1,13 @@
+module.exports = babelJest.createTransformer( {
+	presets: [
+		require.resolve( '@automattic/calypso-build/babel/default' ),
+		require.resolve( '@automattic/calypso-build/babel/wordpress-element' ),
```
The `@automattic/calypso-build` preset provides this transform _without_ this line.

That is something the preset should provide, but I'm not sure of the best way to go about it.

This aligns with the way we pass `--env.WP` to webpack, but Jest doesn't seem to have any concept of arbitrary arguments we pass through the CLI. In fact, jest fails if it doesn't understand the arguments passed to the CLI.

This is sufficient for now, but it's something that should be considered for future calypso-build work.

ockham replied:
Good point, yep :+1:

sirreal committed:
Add @babel/preset-env dep

sirreal committed:
Remove fixed @TODO

This seems to have been fixed by https://github.com/Automattic/wp-calypso/pull/34860

sirreal commented on webpack.config.extensions.js:
```diff
@@ -114,12 +114,6 @@ module.exports = [
 	},
 	{
 		...componentsWebpackConfig,
-		output: {
```
Fixed by https://github.com/Automattic/wp-calypso/pull/34860? @ockham

ockham replied:
Yes! Good spot. Sorry for not cleaning up after myself :sweat_smile:

ockham reviewed (commented):
This looks great and seems to build and work well (both dashboard and Gutenblocks); thanks a ton for working on this!

One note, I see a weird visual change in a Gridicon (in the Upgrade Nudge) :confounded: 

![image](https://user-images.githubusercontent.com/96308/65469840-1f535c80-de69-11e9-8ebe-45177d6d17d1.png)

(Note that the Nudge isn't enabled yet -- I'm enabling it in my dev env through  `add_filter( 'jetpack_block_editor_enable_upgrade_nudge', '__return_true' );`. Probably fine to fix in a follow-up, as long as nothing else is broken -- I couldn't spot anything :sweat_smile: )

simison commented:
>  One note, I see a weird visual change in a Gridicon (in the Upgrade Nudge) 😖

Could this be related to color-studio changes in Calypso build pipeline?

sirreal commented on yarn.lock:
```diff
@@ -41,11 +45,16 @@
   resolved "https://registry.yarnpkg.com/@automattic/calypso-color-schemes/-/calypso-color-schemes-1.0.0.tgz#17a14e3257bd90b40e960d624cca4353d4d20c34"
   integrity sha512-W7r4pgcBauLx65oTLbHKV84ScnfEKmW7T/sXkPfByHhuIE7+OpubmiiBsizWatk1I/puUTmj+SDqEOJMOyRdzA==
 
+"@automattic/calypso-color-schemes@^2.0.2":
```
@ockham This may be related to the visual change you're seeing…
https://github.com/Automattic/jetpack/pull/13503#pullrequestreview-292105012

sirreal replied:
Yeah… I think this line:

https://github.com/Automattic/jetpack/blob/f55f55e67b1071518ece4318a349e359acf88309/extensions/shared/components/upgrade-nudge/style.scss#L6

Would need to become [`--studio-white`](https://github.com/Automattic/color-studio/blob/4b3262ac652a39c434c818e9e9c357fc23c8c0cc/dist/color-properties.css#L4).

simison replied:
@porada :wave: where there any other colors than `white` that now need `--studio` prefix?

Bummer with CSS variables (as opposite to scss variables) is that they fail silently. :-/

Checked if these would be listed on changelog but no dice: 
https://github.com/Automattic/color-studio/blob/4b3262ac652a39c434c818e9e9c357fc23c8c0cc/CHANGELOG.md

porada replied:
We replaced white with 3 semantic properties:

* `--color-surface` (default white background),
* `--color-text-inverted` (white text, which is also the correct substitute for the case mentioned above),
* `--color-border-inverted` (white border).

Generally, we should avoid `--studio-*` values in favor of the [theme properties](https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-color-schemes/src/shared/color-schemes/_default.scss) (that’s a full listing).

---

> Bummer with CSS variables (as opposite to scss variables) is that they fail silently

Yes, that is a disadvantage of using CSS properties. I hope we can make Stylelint throw errors when using an undefined property.

> Checked if these would be listed on changelog but no dice

Color Studio’s changelog focuses only on color value changes. [`calypso-color-schemes`](https://github.com/Automattic/wp-calypso/blob/master/packages/calypso-color-schemes/CHANGELOG.md) would be a better place to look this up.

simison replied:
Gotcha, thanks for a thorough overview @porada !

simison replied:
Opened an issue to separate colors from `calypso-build`: https://github.com/Automattic/wp-calypso/issues/36466

brbrr reviewed (dismissed):
I've done some functional testing, and all looks good. LGTM :shipit:

zinigor commented:
Cherry-picking yarn.lock didn't really fix the conflict, this needs a rebase. Can one of the authors please rebase when you're happy with the reviews, so that the history doesn't get lost?

sirreal committed:
Update yarn.lock

sirreal commented:
Rebased and yarn.lock regenerated

kraftbj reviewed (dismissed):
Tested with a completely clean yarn build and via JN. No errors in CLI, console, and everything seemed functional.

simison committed:
Fix colors

- Switch Muriel `--color-white` to Gutenberg white
- Switch two specific Muriel colors in Mailchimp to Muriel success+error colors like elsewhere in `/extensions/*`

simison commented:
Aight! I tested several places where CSS variables (`--color-white` etc) are used in `extensions` and fixed what was broken.

Instead of switching it to `--color-surface` I switched to Gutenberg `$white` color variable which is:
- Easy to understand since we don't have Calypso color templates yet in the block editor
- We're in Gutenberg context after all, let's use Gutenberg core colors
- It's already used all across, so this is more consistent. I'd expect all SASS variables eventually be migrated to CSS variables but it has to be more controlled process.
- It won't silently break in future updates like changes in CSS color variables do; I'd expect colors to be separated from the build tool so that future updates like this wouldn't be so scary: https://github.com/Automattic/wp-calypso/issues/36466

There are still muriel color variables around these files but those are harder to replace as it would require changing the color entirely.

jeherve reviewed (approved):
This seems to work well for me. 👍

simison commented:
Deployed in r197337-wpcom
